### DISCUSSION TITLE
Should we allow for passing compiled regex strings?

### DISCUSSION THREAD

**Discussion by mdjastrzebski** [callstack/ts-regex-builder#14] [Ideas]
How should we handle cases when string value passed by the user to the builder contains special Regex meaning, e.g.,
```
oneOrMore(".") // in regex "." is special char meaning any
oneOrMore("[a-z]") // [a-z]+
```

We have to options here:
* beginner friendly: escape all passed strings, e.g.,
```
buildRegex(
   oneOrMore(word),
   zeroOrMore(
     ".", // <--- Here is it only dot character OR is it regex special character for any
     choiceOf("txt", "md"),
  )
)
```
* power-user friendly: to allow passing custom compiled regex values, e.g.
```
buildRegex(
  "\w+", 
  zeroOrMore(
    "\.",
    "(txt|md)
  )
)
```

We should establish Swift RB behavior in this regard, as that would be important factor.

Personally I would lean toward "beginner-friendly" as we strive to make Regex easy to use, and allowing for regex special chars, could result in hard to diagnose errors, that would be surprise for users new to regex.

@okwasniewski, @jaworek wdyt?

**Comment by jaworek** [comment]
I also lean into "beginner-friendly" approach. Maybe we can add `customChar` or something that allows user creating their own rule?

**Reply by mdjastrzebski** [reply]
Escape hatch! I like it!
```
buildRegex(
  raw("\w+")
)
```

Let's scout Swift RB to see if they have anything like that.

**Reply by okwasniewski** [reply]
For me also +1 for `raw()` but let's focus on beginners as the default

**Comment by mdjastrzebski** [comment]
Yeah, I'll write it down in some project assumptions doc, somewhere...

**Comment by jaworek** [comment]
Example from Swift RegexBuilder shows that they're escaping "." that user passed when trying to validate regex, so it confirms our approach with "beginner-friendly".
https://developer.apple.com/documentation/regexbuilder#Overview
